File: core/governance/TQM_PolicyDispatcher.py
import json
from typing import Dict, Any

class TQM_PolicyDispatcher:
    def __init__(self, policy_config_path: str):
        self.config = self._load_config(policy_config_path)
        self.gates = self.config.get("quality_gates", {})

    def _load_config(self, path: str) -> Dict[str, Any]:
        with open(path, 'r') as f:
            return json.load(f)

    def evaluate_metric(self, metric_key: str, value: float, context: Dict[str, Any]):
        violations = []
        
        # Iterate through all gates and policies
        for gate_name, gate_data in self.gates.items():
            for policy in gate_data.get('policies', []):
                if policy['metric_key'] == metric_key:
                    violation = self._check_violation(policy, value)
                    if violation:
                        violations.append({
                            "policy_id": policy['id'],
                            "severity": policy['severity'],
                            "handler": policy['enforcement_handler'],
                            "auto_fixable": policy['auto_fixable'],
                            "details": violation
                        })
        
        return violations

    def _check_violation(self, policy: Dict[str, Any], value: float) -> str or None:
        threshold = policy['threshold']
        mode = policy['mode']
        
        if mode == 'MAX_VIOLATION' and value > threshold:
            return f"Value {value} exceeds max threshold {threshold}."
        if mode == 'MIN_REQUIREMENT' and value < threshold:
            return f"Value {value} is below minimum requirement {threshold}."
        
        # Add other modes (e.g., RANGE_VIOLATION)
        
        return None

    def dispatch_violations(self, violations: list):
        # Logic to route violation data to the specified handler (e.g., EvolutionEngine, ArchitectureReviewer)
        for violation in violations:
            handler = violation['handler']
            # Example: A mechanism to invoke handler dynamically
            print(f"Dispatching {violation['severity']} violation to handler: {handler}")
            # Execute: GovernanceKernel.invoke_module(handler, violation)

# Note: Execution logic for specific handlers must be handled by the primary AGI kernel.

File: governance/TQM_PolicyDispatcher.js
import fs from 'fs';

/**
 * TQM_PolicyDispatcher handles the evaluation of metrics against defined quality gates
 * and dispatches resulting violations to appropriate enforcement handlers.
 */
class TQM_PolicyDispatcher {
    /**
     * @param {string} policyConfigPath - Path to the JSON configuration file defining policies and gates.
     */
    constructor(policyConfigPath) {
        // Note: Synchronous loading is used for initialization, typical for configuration modules.
        this.config = this._loadConfig(policyConfigPath);
        this.gates = this.config.quality_gates || {};
    }

    /**
     * Loads the policy configuration synchronously.
     * @param {string} filePath
     * @returns {object}
     * @private
     */
    _loadConfig(filePath) {
        try {
            const rawData = fs.readFileSync(filePath, 'utf8');
            return JSON.parse(rawData);
        } catch (error) {
            console.error(`[TQM Dispatcher] Failed to load configuration from ${filePath}:`, error.message);
            return { quality_gates: {} };
        }
    }

    /**
     * Evaluates a single metric value against all relevant policies.
     * @param {string} metricKey - The identifier of the metric being checked.
     * @param {number} value - The numerical value of the metric.
     * @param {object} context - Additional context data.
     * @returns {Array<object>} List of violations found.
     */
    evaluateMetric(metricKey, value, context = {}) {
        const violations = [];
        
        // Iterate through all gates and policies
        for (const gateName in this.gates) {
            const gateData = this.gates[gateName];
            const policies = gateData.policies || [];

            for (const policy of policies) {
                if (policy.metric_key === metricKey) {
                    const violationDetails = this._checkViolation(policy, value);
                    if (violationDetails) {
                        violations.push({
                            policy_id: policy.id,
                            severity: policy.severity,
                            handler: policy.enforcement_handler,
                            auto_fixable: policy.auto_fixable,
                            details: violationDetails
                        });
                    }
                }
            }
        }
        
        return violations;
    }

    /**
     * Checks if a specific value violates a policy threshold.
     * @param {object} policy
     * @param {number} value
     * @returns {string | null} Violation message or null if no violation.
     * @private
     */
    _checkViolation(policy, value) {
        const threshold = policy.threshold;
        const mode = policy.mode;
        
        if (mode === 'MAX_VIOLATION' && value > threshold) {
            return `Value ${value} exceeds max threshold ${threshold}.`;
        }
        if (mode === 'MIN_REQUIREMENT' && value < threshold) {
            return `Value ${value} is below minimum requirement ${threshold}.`;
        }
        
        return null;
    }

    /**
     * Dispatches violations to their specified handlers.
     * @param {Array<object>} violations
     */
    dispatchViolations(violations) {
        for (const violation of violations) {
            const handler = violation.handler;
            console.log(`[Dispatcher] Dispatching ${violation.severity} violation to handler: ${handler}`);
            // Execution: UNIFIER.invokeModule(handler, violation);
        }
    }
}

export { TQM_PolicyDispatcher };